pkg/services/accesscontrol/filter.go

```diff
@@ -0,0 +1,96 @@
+package accesscontrol
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+)
+
+var sqlIDAcceptList = map[string]struct{}{}
+
+type SQLDialect interface {
+	DriverName() string
+}
+
+// Filter creates a where clause to restrict the view of a query based on a users permissions
+// Scopes for a certain action will be compared against prefix:id:sqlID where prefix is the scope prefix and sqlID
+// is the id to generate scope from e.g. user.id
+func Filter(ctx context.Context, dialect SQLDialect, sqlID, prefix, action string, user *models.SignedInUser) (string, []interface{}, error) {
+	if _, ok := sqlIDAcceptList[sqlID]; !ok {
+		return "", nil, errors.New("sqlID is not in the accept list")
+	}
+
+	var scopes []string
+	if user.Permissions != nil && user.Permissions[user.OrgId] != nil {
+		scopes = append(scopes, user.Permissions[user.OrgId][action]...)
+	}
+
+	if len(scopes) == 0 {
+		scopes = append(scopes, "no:access")
+	}
+
+	var sql string
+	var args []interface{}
+
+	if strings.Contains(dialect.DriverName(), migrator.SQLite) {
+		sql, args = sqliteQuery(scopes, sqlID, prefix)
+	} else if strings.Contains(dialect.DriverName(), migrator.MySQL) {
+		sql, args = mysqlQuery(scopes, sqlID, prefix)
+	} else {
+		sql, args = postgresQuery(scopes, sqlID, prefix)
+	}
```

Comment:
Two suggestions:

- Use switch instead of if to make it clear that this is a range of equivalent options.
- Default to error to make sure the error doesn't happen in SQL down the line.

```suggestion
	switch {
	case strings.Contains(dialect.DriverName(), migrator.SQLite):
		sql, args = sqliteQuery(scopes, sqlID, prefix)
	case strings.Contains(dialect.DriverName(), migrator.MySQL):
		sql, args = mysqlQuery(scopes, sqlID, prefix)
	case strings.Contains(dialect.DriverName(), migrator.Postgres)
		sql, args = postgresQuery(scopes, sqlID, prefix)
	default:
		return nil, nil, fmt.Errorf("unknown database: %s", dialect.DriverName())
	}
```